### PR TITLE
warn user when dropping unpicklable hparams

### LINKED ISSUE #2444: self.hparam silently removes params that are not serializable
## 🐛 Bug

Following the approach found under [hyperparameters in the docs](https://pytorch-lightning.readthedocs.io/en/latest/hyperparameters.html#lightningmodule-hyperparameters), step 3, I passed a dict with parameters to my `pl.LightningModule`.
In the `__init__` printing `self.hparams` shows all contents I passed.
However, in the function `def configure_optimizers(self)`, some hparams are gone.

This might be related to that not all param values are YAML serializable, and therefore automatically removed? Because the 2 removed params are `"criterion": torch.nn.BCELoss()` and `"optimizer": partial(optim.Adam, lr=0.001)`.

### To Reproduce

1. Run the following script:
```python
from functools import partial
import torch
import torch.optim as optim
from torch.utils.data import Dataset
import pytorch_lightning as pl
from pytorch_lightning import Trainer

# partial to give all params, except the data
hparams = {
    "criterion": torch.nn.BCELoss(),  # F.cross_entropy(),  # loss function
    "optimizer": partial(optim.Adam, lr=0.001),  # (lr=0.001),
    # "learning_rate": 0.001,
    "filters": 64,
    "layers": 2
}

class EmptyDataset(Dataset):
    def __init__(self, transform=None):
        pass
    
    def __len__(self):
        return 32
    
    def __getitem__(self, idx):
        return {"input": np.array([1]), "output": "nothing"}

class LitLake(pl.LightningModule):
    def __init__(self, hparams: dict, transforms: dict = None):
        super().__init__()
        
        self.hparams = hparams
        print("self.hparams\n", self.hparams)
        
    def forward(self, x):
        pass
    
    def training_step(self, batch, batch_idx):
        """
        Lightning calls this inside the training loop with the data from the training dataloader
        passed in as `batch`.
        """
        # forward pass
        x, y = batch
        y_hat = self(x)
        loss = self.hparams["criterion"](y_hat, y)
        tensorboard_logs = {'train_loss': loss}
        return {'loss': loss, 'log': tensorboard_logs}
        
    def configure_optimizers(self):
        print("self.hparams\n", self.hparams)
        optimizer = self.hparams["optimizer"](self.parameters())
        scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=10)
        return [optimizer], [scheduler]
    
    def train_dataloader(self):
         return DataLoader(EmptyDataset(), batch_size=4, num_workers=1)

model = LitLake(hparams=hparams)
# most basic trainer, uses good defaults
trainer = Trainer()  # gpus=1, num_nodes=1
trainer.fit(model)  # KeyError: 'optimizer'
```
2. See error

<details><summary>Script output (CLICK ME)</summary>
<p>

```python
self.hparams
 "criterion": BCELoss()
"filters":   64
"layers":    2
"optimizer": functools.partial(<class 'torch.optim.adam.Adam'>, lr=0.001)
GPU available: True, used: False
TPU available: False, using: 0 TPU cores
self.hparams
 "filters": 64
"layers":  2
Traceback (most recent call last):
  File "lightning_hparams_bug.py", line 61, in <module>
    trainer.fit(model)  # KeyError: 'optimizer'
  File "/home/*user*/anaconda3/envs/onseilake/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 965, in fit
    self.optimizers, self.lr_schedulers, self.optimizer_frequencies = self.init_optimizers(model)
  File "/home/*user*/anaconda3/envs/onseilake/lib/python3.7/site-packages/pytorch_lightning/trainer/optimizers.py", line 18, in init_optimizers
    optim_conf = model.configure_optimizers()
  File "lightning_hparams_bug.py", line 51, in configure_optimizers
    optimizer = self.hparams["optimizer"](self.parameters())
KeyError: 'optimizer'
```

</p>
</details>

### Expected behavior

Either:
1. `self.hparams` keeping the non-serializable parameters (will give problems with loading?)
2. Throw an error explaining why those param values are not acceptable and how to approach it, instead of silently removing them.

### Environment

 - PyTorch Version (e.g., 1.0): 1.5.0
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): `conda`
 - Build command you used (if compiling from source): -
 - Python version: 3.7.6
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: 1x GeForce GTX 1080 Ti
 - Any other relevant information: -

### Additional context

Bug reproduced by @Borda

### LINKED ISSUE #2676: Namespace Cleaning
## 🚀 Feature
Change how the namespace is cleaned to be fully general. Right now, it filters all callables, but this filters some things which are pickable and callable.

### Motivation
Currently, users can interact with callable objects only in the init method. It's useful to  use these elsewhere, for example I pass in optim.Adam or optim.SGD to switch between optimizers, more flexible than hardcoding strings.

### Pitch

rewrite clean_namespace to be something like:

```
for k,v in hparams.items():
    try:
        pickle.dumps(v)
    except:
        del_attrs.append(k)
```

### PR DESCRIPTION
**Pull request by monney**
## What does this PR do?
PR changes namespace cleaning to be more general. Makes sure it only removes items which can't be pickled, and warns user when doing so.

Fixes #2676 
Fixes #2444

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
@monney I updated your pr descripion. Since you added a warning, I believe this also solves a the issue #2444.

**Comment by awaelchli** [comment]
@monney is this wip or ready to review? at this point I would suggest to extract a helper function `is_pickleable(obj: object)` to avoid duplicated code as it is useful in other places too. Thanks for your efforts.

**Comment by monney** [comment]
@awaelchli Should be good to go now. Build is failing, but I believe the test case is broken.

**Comment by monney** [comment]
@awaelchli I added your test case and changed the warning to rank_zero_warn, I think everything should be good to go now!

**Comment by awaelchli** [comment]
@monney There are test failures, they must be because of the new way of cleaning the namespace, but not sure what needs to be changed. Have you seen it?

**Comment by monney** [comment]
@awaelchli it is because of the new way of cleaning. Something prevented that test case either from being run or failing previously. The property that it’s checking for would have been completely removed by the old way of cleaning, I’m not sure why that did not trigger a failure though. The new way causes it to fail because it should be checking for CosineEmbedding.

**Comment by awaelchli** [comment]
Ah I thought I already confirmed it. Yes, please change to CosineEmbeddingLoss I agree.

**Comment by monney** [comment]
You did. That’s my bad. Doesn’t it need to be changed in the CI server? Or if I change it in my code will it update

**Comment by awaelchli** [comment]
It's in the same file you added the test, above in a fuction called test_collect_init_arguments. You change it there, push your commit and it will show here in the tab for changed files :)

**Comment by monney** [comment]
I got it now, I didn't realize it ran the test cases, from the commit itself, and thought you had update it elsewhere. Thank you! :)

**Review by awaelchli — APPROVED** [review]
@monney thanks for the fix

File: pytorch_lightning/utilities/parsing.py [hunk 1]
```diff
@@ -25,26 +28,28 @@ def str_to_bool(val):
...
-        del_attrs = []
-        for k, v in hparams.items():
-            if callable(v):
-                del_attrs.append(k)
-
-        for k in del_attrs:
-            del hparams[k]
+        hparams_dict = hparams.__dict__
+
+    del_attrs = [k for k, v in hparams_dict.items() if not is_picklable(v)]
```
Comment (nateraw) [inline pytorch_lightning/utilities/parsing.py:48]: nice

**Review by SkafteNicki — APPROVED** [review]
LGTM :+1:

File: pytorch_lightning/utilities/parsing.py [hunk 2]
```diff
@@ -25,26 +28,28 @@ def str_to_bool(val):
         raise ValueError(f'invalid truth value {val}')
 
 
+def is_picklable(obj: object) -> bool:
```
Comment (Borda) [inline pytorch_lightning/utilities/parsing.py:31]: this can be quite a time costly especially for large objects or calling this function quite often...

**Pull request merged**